File: code/jvm/im-services/src/main/kotlin/im/services/messages/MessageService.kt
package im.services.messages

import im.domain.Either
import im.domain.messages.Message
import im.domain.user.User
import im.domain.wrappers.identifier.Identifier
import im.repository.pagination.Pagination
import im.repository.pagination.PaginationRequest
import im.repository.pagination.SortRequest
import java.time.LocalDateTime

interface MessageService {
    /**
     * Retrieves the messages of a channel.
     *
     * A user must be a member of the channel to retrieve its messages.
     *
     * @param channelId the unique identifier of the channel
     * @param pagination the pagination settings
     * @param user the user that is requesting the messages
     * @param sortRequest the sort settings
     * @return a list of messages
     */
    fun getChannelMessages(
        channelId: Identifier,
        pagination: PaginationRequest,
        sortRequest: SortRequest,
        user: User,
    ): Either<MessageError, Pagination<Message>>

    /**
     * Creates a message in a channel.
     *
     * A user must be a member of the channel and have write permission to create a message.
     *
     * @param channelId the unique identifier of the channel
     * @param message the content of the message
     * @param user the user that is creating the message
     * @return the created message
     */
    fun createMessage(
        channelId: Identifier,
        message: String,
        user: User,
    ): Either<MessageError, Message>

    /**
     * Updates the content of a message.
     *
     * A user must be the author of the message to update its content.
     *
     * @param channelId the unique identifier of the channel
     * @param messageId the unique identifier of the message
     * @param message the new content of the message
     * @param user the user that is updating the message
     * @return the date and time when the message was updated
     */
    fun updateMessage(
        channelId: Identifier,
        messageId: Identifier,
        message: String,
        user: User,
    ): Either<MessageError, LocalDateTime>

    /**
     * Deletes a message from a channel.
     *
     * A user must be the author of the message or the owner of the channel to delete a message.
     *
     * @param channelId the unique identifier of the channel
     * @param messageId the unique identifier of the message
     * @param user the user that is deleting the message
     * @return a unit value
     */
    fun deleteMessage(
        channelId: Identifier,
        messageId: Identifier,
        user: User,
    ): Either<MessageError, Unit>

    /**
     * Retrieves a message by its unique identifier.
     *
     * A user must be a member of the channel to retrieve the message.
     *
     * @param channelId the unique identifier of the channel
     * @param messageId the unique identifier of the message
     * @param user the user that is requesting the message
     * @return the message
     */
    fun getMessageById(
        channelId: Identifier,
        messageId: Identifier,
        user: User,
    ): Either<MessageError, Message>
}

File: code/js/src/Services/messages/MessageService.ts
import { Channel } from '../../Domain/channel/Channel';
import { PaginationRequest } from '../../Domain/pagination/PaginationRequest';
import { SortRequest } from '../../Domain/pagination/SortRequest';
import { ApiResult, handle } from '../media/Problem';
import { MessagesPaginatedOutputModel } from '../../Dto/output/messages/MessagesPaginatedOutputModel';
import { buildQuery } from '../Utils';
import { BaseHTTPService } from '../BaseHTTPService';
import { Uri } from '../Uri';
import { Pagination } from '../../Domain/pagination/Pagination';
import { Message } from '../../Domain/messages/Message';
import { Identifier } from '../../Domain/wrappers/identifier/Identifier';
import { MessageOutputModel } from '../../Dto/output/messages/MessageOutputModel';
import { MessageCreationInputModel } from '../../Dto/input/MessageCreationInputModel';
import { MessageCreationOutputModel } from '../../Dto/output/messages/MessageCreationOutputModel';
import { Session } from '../../Domain/sessions/Session';

export namespace MessageService {
    import get = BaseHTTPService.get;
    import CHANNEL_ID_PARAM = Uri.CHANNEL_ID_PARAM;
    import MESSAGES_ROUTE = Uri.MESSAGES_ROUTE;
    import MESSAGE_ROUTE = Uri.MESSAGE_ROUTE;
    import MESSAGE_ID_PARAM = Uri.MESSAGE_ID_PARAM;
    import deleteRequest = BaseHTTPService.deleteRequest;
    import post = BaseHTTPService.post;
    import put = BaseHTTPService.put;

    /**
     * Gets the messages of a channel.
     *
     * @param channel - The channel to get the messages from.
     * @param pagination - The pagination request.
     * @param sort - The sort request.
     * @param before - The date to get messages before.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The messages of the channel.
     */
    export async function getChannelMessages(
        channel: Channel,
        pagination?: PaginationRequest,
        sort?: SortRequest,
        before?: Date,
        abortSignal?: AbortSignal,
    ): ApiResult<Pagination<Message>> {
        return await handle(
            get<MessagesPaginatedOutputModel>({
                uri: buildQuery(
                    MESSAGES_ROUTE,
                    null,
                    pagination,
                    sort,
                    false,
                    null,
                    before,
                ).replace(CHANNEL_ID_PARAM, channel.id.value.toString()),
                abortSignal: abortSignal,
            }),
            (outputModel) => ({
                items: outputModel.messages.map(Message.fromDto),
                info: outputModel.pagination,
            }),
        );
    }

    /**
     * Gets a message by its identifier.
     *
     * @param channel - The channel the message belongs to.
     * @param messageId - The identifier of the message.
     * @param abortSignal - The signal to abort the request.
     *
     * @returns The message.
     */
    export async function getMessage(
        channel: Channel,
        messageId: Identifier,
        abortSignal?: AbortSignal,
    ): ApiResult<Message> {
        return await handle(
            get<MessageOutputModel>({
                uri: MESSAGE_ROUTE.replace(
                    CHANNEL_ID_PARAM,
                    channel.id.value.toString(),
                ).replace(MESSAGE_ID_PARAM, messageId.value.toString()),
                abortSignal: abortSignal,
            }),
            (outputModel) => Message.fromDto(outputModel),
        );
    }

    /**
     * Creates a message.
     *
     * @param channel - The channel to create the message in.
     * @param content - The content of the message.
     * @param session - The session of the user creating the message.
     * @param abortSignal - The signal to abort the request.
     */
    export async function createMessage(
        channel: Channel,
        content: string,
        session: Session,
        abortSignal?: AbortSignal,
    ): ApiResult<Message> {
        return await handle(
            post<MessageCreationInputModel, MessageCreationOutputModel>({
                uri: MESSAGES_ROUTE.replace(
                    CHANNEL_ID_PARAM,
                    channel.id.value.toString(),
                ),
                requestBody: {
                    content: content,
                },
                abortSignal: abortSignal,
            }),
            (outputModel) =>
                Message.fromCreation(
                    outputModel,
                    channel.id,
                    session.user,
                    content,
                ),
        );
    }

    /**
     * Updates a message.
     *
     * @param message - The message to update.
     * @param content - The new content of the message.
     * @param abortSignal - The signal to abort the request.
     */
    export async function updateMessage(
        message: Message,
        content: string,
        abortSignal?: AbortSignal,
    ): ApiResult<void> {
        return await put<MessageCreationInputModel>({
            uri: MESSAGE_ROUTE.replace(
                CHANNEL_ID_PARAM,
                message.channelId.value.toString(),
            ).replace(MESSAGE_ID_PARAM, message.id.value.toString()),
            requestBody: {
                content: content,
            },
            fetchResBody: false,
            abortSignal: abortSignal,
        });
    }

    /**
     * Deletes a message.
     *
     * @param channel - The channel the message belongs to.
     * @param messageId - The identifier of the message.
     */
    export async function deleteMessage(
        channel: Channel,
        messageId: Identifier,
    ): ApiResult<void> {
        return await deleteRequest({
            uri: MESSAGE_ROUTE.replace(
                CHANNEL_ID_PARAM,
                channel.id.value.toString(),
            ).replace(MESSAGE_ID_PARAM, messageId.value.toString()),
        });
    }
}
